packages/alfa-wcag/test/helpers/outcome.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  Aspect,
+  AspectsFor,
+  audit,
+  isResult,
+  Outcome,
+  Rule,
+  Target
+} from "@siteimprove/alfa-act";
+import { Assertions } from "@siteimprove/alfa-test";
+import { concat } from "@siteimprove/alfa-util";
+
+export function outcome<A extends Aspect, T extends Target>(
+  t: Assertions,
+  rule: Rule<A, T>,
+  aspects: AspectsFor<A>,
+  assert:
+    | Outcome.Inapplicable
+    | Readonly<
+        { [O in Outcome.Failed | Outcome.Passed | Outcome.CantTell]?: Array<T> }
+      >,
+  dependencies: Array<Rule<A, T>> = []
+) {
+  const outcomes: Array<Outcome.Passed | Outcome.Failed | Outcome.CantTell> = [
+    Outcome.Passed,
+    Outcome.Failed,
+    Outcome.CantTell
+  ];
+
+  const results = audit(aspects, concat([rule], dependencies)).filter(
+    result => result.rule === rule.id
+  );
+
+  if (assert === Outcome.Inapplicable) {
+    t.equal(results.length, 1, "There must only be one result");
+    const result = results[0];
+    t(
+      isResult(result) && result.outcome === Outcome.Inapplicable,
+      "The outcome must be inapplicable"
+    );
+    return;
+  }
+
+  for (const outcome of outcomes) {
+    const actual = results
+      .filter(isResult)
+      .filter(result => result.outcome === outcome);
+
+    const expected = outcome in assert ? assert[outcome]! : [];
+
+    t.equal(
+      actual.length,
+      expected.length,
+      `There must be ${expected.length} ${outcome} results`
+    );
+
+    for (const target of expected) {
+      if (target === null) {
+        // Black magic will happen here
+      } else {
+        const holds = actual.some(
+          result =>
+            result.outcome === Outcome.Inapplicable || result.target === target
```

Comment:
Inapplicable results should be handled by previous code so try removing the first condition and see if the code still types.

Reply:
Unfortuntately, removing the Inapplicable part results in a compile-error.
```
Property 'target' does not exist on type 'Readonly<{ rule: string; outcome: Outcome.Inapplicable; }> | Readonly<{ rule: string; outcome: Outcome.Passed | Outcome.Failed; target: T; }>'.
  Property 'target' does not exist on type 'Readonly<{ rule: string; outcome: Outcome.Inapplicable; }>'
```

Reply:
I'm guessing that's because this is not a proper type guard: https://github.com/Siteimprove/alfa/pull/82/files#diff-34397abfa06ec59f4dac72d55a3c25fcR47

Reply:
Do you have any suggestions on a way to implement this type-guard?

Reply:
Extracting the lambda and annotating it as a type guard should do the trick: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

Reply:
I tried exactly that, but I only managed to construct the inverse type-guard.

```ts
export function isInapplicable(
  input: Result
): input is Readonly<{ rule:string; outcome: Outcome.Inapplicable }> {
  return input.outcome === Outcome.Inapplicable;
}
``` 
works as expected, while 
```ts
export function isNotInapplicable(
  input: Result
): input is Readonly<{ rule:string; outcome: Outcome.Failed | Outcome.Passed; target: Target }> {
  return input.outcome === Outcome.Failed || input.outcome === Outcome.Passed;
}
```
does not change the end type at all.

Reply:
Results are now generic in their outcome, which made the following type guard possible:

```ts
function hasMatchingOutcome(result: Result): result is Result<T, typeof outcome> {
  return result.outcome === outcome;
}
```

Reply:
I don't think that would work, as Result only takes one argument (T).
```ts 
Result<T, typeof outcome>
```

Reply:
f0437fb95c044686de2eeda94d9e6e9555e96cf7 😉

Reply:
Right, thanks, I think it is good to go then